--- a/trivium.c	2006-02-24 11:16:21.000000000 +1300
+++ b/trivium.c	2011-04-17 13:36:04.214875003 +1200
@@ -313,4 +313,35 @@
   EMPTY();
 }
 
+/* only deal with multiples of 16 */
+void ECRYPT_keystream_bytes(
+  ECRYPT_ctx* ctx,
+  u8* output,
+  u32 len)
+{
+  m64 s11, s12;
+  m64 s21, s22;
+  m64 s31, s32;
+
+  LOAD(ctx->state);
+
+  for (; (int32_t) len > 0; output += 16, len -= 16)
+    {
+      m64 t1, t2, t3, z[2];
+      UPDATE(); z[0] = XOR(XOR(s12, s22), s32); ROTATE();
+      UPDATE(); z[1] = XOR(XOR(s12, s22), s32); ROTATE();
+
+      M64TO64_CONVERT(z[0]); ((m64*)output)[0] = z[0];
+      M64TO64_CONVERT(z[1]); ((m64*)output)[1] = z[1];
+    }
+  STORE(ctx->state);
+
+  EMPTY();
+}
+
+
 /* ------------------------------------------------------------------------- */
