--- chacha8-orig.c	2011-04-30 20:51:15.444629013 +1200
+++ chacha8.c	2011-04-30 20:55:21.604628994 +1200
@@ -80,7 +80,7 @@
 void ECRYPT_encrypt_bytes(ECRYPT_ctx *x,const u8 *m,u8 *c,u32 bytes)
 {
   u8 output[64];
-  int i;
+  unsigned int i;
 
   if (!bytes) return;
   for (;;) {
@@ -106,9 +106,26 @@
   ECRYPT_encrypt_bytes(x,c,m,bytes);
 }
 
+#if 0
 void ECRYPT_keystream_bytes(ECRYPT_ctx *x,u8 *stream,u32 bytes)
 {
   u32 i;
   for (i = 0;i < bytes;++i) stream[i] = 0;
   ECRYPT_encrypt_bytes(x,stream,stream,bytes);
 }
+#else
+void ECRYPT_keystream_bytes(ECRYPT_ctx *x, u8 *stream, u32 bytes)
+{
+  if (!bytes) return;
+  for (;bytes >= 64;) {
+    salsa20_wordtobyte(stream, x->input);
+    x->input[8] = PLUSONE(x->input[8]);
+    if (!x->input[8]) {
+      x->input[9] = PLUSONE(x->input[9]);
+      /* stopping at 2^70 bytes per nonce is user's responsibility */
+    }
+    stream += 64;
+    bytes -= 64;
+  }
+}
+#endif
